### ISSUE TITLE
Porting Apache NimBLE

### ISSUE THREAD

**Issue by dcz-self** [tock/tock#2970]
The topic of Bluetooth has been taken up in other issues, but it seems there's still little research done on the topic.

Instead of polluting the [smartwatch](https://github.com/tock/tock/issues/2846#issuecomment-1032699337) issue, I decided to start a new one specifically to make notes about NimBLE.

The [slide deck](https://summit.riot-os.org/2018/wp-content/uploads/sites/10/2018/09/1_1-Szymon-Janc-NimBLE.pdf) as well as the [readme](https://github.com/apache/mynewt-nimble#browsing) show the general architecture, which seems to be sliced about right for use in Tock's userspace:

> - nimble/[controller](https://github.com/apache/mynewt-nimble/tree/master/nimble/controller): Contains code for controller including Link Layer and HCI implementation (controller)
> - nimble/[drivers](https://github.com/apache/mynewt-nimble/tree/master/nimble/drivers): Contains drivers for supported radio transceivers (Nordic nRF51 and nRF52) (drivers)
> - nimble/host: Contains code for host subsystem.

NimBLE on Tock was [called possible](https://github.com/tock/tock/issues/2846#issuecomment-1032728713) if
> NimBLE has an abstraction layer which allows routing accesses to the radio peripheral through the kernel

and this seems to be the "drivers" layer. Namely, for the nRF52840, it would be those files that need to get rewritten: https://github.com/apache/mynewt-nimble/tree/master/nimble/drivers/nrf5340/src

From a quick scan, the following functionality is used there:
- RADIO hardware,
- FICR — Factory information configuration registers
- ECB, CCM — AES encryption
- RNG
- AAR — Accelerated address resolver for BLE
- timer and RTC
- GPIO
- PPI — Programmable peripheral interconnect

In total, that's about 2500 lines of C.

One way of approaching a port would be to give NimBLE direct access to the hardware, and then gradually route small groups of functions through the kernel.

**Comment by alexandruradovici** [comment]
From what I understand here, to use NimBLE as an application, it would require access to all these listed peripherals?

I think giving applications direct access to hardware is not something that we would like to do in Tock. I think most of these peripherals already have drivers that allow applications to access them. What about having a driver that would simply expose the Radio (send / receive bytes and some configuration) hardware functions? Would that work?

**Comment by dcz-self** [comment]
> to use NimBLE as an application, it would require access to all these listed peripherals?

Yes, and no.

Yes, in that NimBLE needs to have access to the functionality. No, because that access can be mediated by the kernel (at least I didn't notice anything preventing that so far). Yes, because giving the application direct access to the hardware is IMO the most practical way to get the syscall layer off the ground (as described in my original post).

> What about having a driver that would simply expose the Radio (send / receive bytes and some configuration) hardware functions? Would that work?

Yes, that's precisely what I have in mind. Might start dabbling with this next week.

**Comment by alexandruradovici** [comment]
As long as this does not imply an application having direct access to hardware, I am eager to see a working BLE driver in Tock.

**Comment by alistair23** [comment]
> One way of approaching a port would be to give NimBLE direct access to the hardware, and then gradually route small groups of functions through the kernel.

I'm not convinced that will work unfortunately.

I have tried something similar with the Apollo3 BLE stack before. I had it working in userspace by just disabling memory protections and letting the library directly access the Bluetooth hardware. With the goal to them convert that to use syscalls. Unfortunately Bluetooth is very time sensitive and I never managed to get the Bluetooth stack to work when routing the data through Tock syscalls.

I suspect BLE is only possible when it is in the kernel due to the syscall overhead

**Comment by dcz-self** [comment]
Do you still have the code from that experiment?

How big is the overhead of syscalls in Tock? Do syscalls cause rescheduling?

How could the problem get mitigated? Is it a matter of choosing the right abstraction level at which to place the syscalls?

**Comment by alexandruradovici** [comment]
> How big is the overhead of syscalls in Tock? Do syscalls cause rescheduling?
> 
We did test that for the micro:bit, a system call round trip seems to be around 5127 μs for command to command and 9213 μs for command - subscribe - upcall. I can send you the full report is interested, send me an email at alexandru.radovici at wyliodrin.com

System call do cause rescheduling, meaning that after the execution of the system call (not necessarily the action requested  by the system call, the action could be asynchronous), the scheduler is called to choose a process. This might be the same process that issued the system call or another one. On the bright side, you can write your own scheduler that takes into account this.

> How could the problem get mitigated? Is it a matter of choosing the right abstraction level at which to place the syscalls?

Maybe you could write *half* of the BLE driver in user space and the time sensitive part in a capsule.

**Comment by dcz-self** [comment]
Oh wow, those numbers are atrocious. For short calls, the majority of the time is going to be spent in ovrehead. I do hope those values come from another process getting run in between, because then at least it's possible to optimize out.

**Comment by alexandruradovici** [comment]
We did test several use cases, those are numbers (I think for Tock 1.6) when no other process is running. It gets worse if other processes are running.

**Comment by alistair23** [comment]
> Do you still have the code from that experiment?

Yeah, it's pretty messy and getting old now, but you can see it:
https://github.com/alistair23/tock/commits/alistair/apollo3-ble-us
https://github.com/alistair23/libtock-c/commits/alistair/apollo3-ble

This is the change to replace the direct call to hardware with a syscall: https://github.com/alistair23/libtock-c/commit/43a58b368b3901fb733ce63e98e625a3cf0a2e35#diff-47d045f08b745174e90b6a62dd90c3420a6391c16cb3763184c7e943bff1a21bR1249

From memory I checked and we did the exact same thing in Tock, just with a large overhead

**Comment by hudson-ayers** [comment]
Regarding system call overhead, we had RISC-V round-trip syscall overhead measured at 1100 *cycles* in mid-2020 (https://github.com/tock/tock/issues/1730), which would be 17 microseconds @ 64 MHz. If that number has increased by two orders of magnitude there is probably something wrong that we need to investigate.

**Comment by alistair23** [comment]
Yeah, you can run everything in userspace with the memory protection gone. I was trying to move the hardware access parts into the kernel and leave the logic in userspace.

The idea being that the kernel read/writes the hardware registers, but we run the complex BLE stack in userspace so that it's isolated. Plus it meant I could start from a known working implementation. Unfortunately I never got the hardware accesses to work in the kernel

**Comment by psyspy1** [comment]
@dcz-self Hello! Any updates on this effort? I would love to see BLE support in Tock (and eventually in OpenSK). I am a newbie in the embedded world but interested in learning and contribution. Please let me know how I could with this.

**Comment by dcz-self** [comment]
I made some progress analyzing what NimBLE needs,but no code yet. If you want to help, I would be grateful if you found a way to make it compile as part of libtock-c. The documentation for porting is somewhat lacking.

**Comment by alistair23** [comment]
I had another look at this, and here is my current thinking of the best option.

### Run Nimble in the kernel

We can include nimble in the kernel. This has the large downside of including C code in the kernel, but it really seems like the only way Tock will ever get BLE support. It should be behind a compile time option/flag so it can be disabled if it isn't required.

This gets us a fully certified BLE stack maintained by the Apache foundation. I hope it could be something that can be upstreamed

There is already a few other Nimble ports that can be used as examples

- https://github.com/taks/esp32-nimble
- https://github.com/h2zero/NimBLE-Arduino

The [cc](https://docs.rs/cc/latest/cc/) crate could be used at build time to pull in Nimble. Then we need to hide the Nimble HCI API behind a Tock capsule.

### Porting

We would need to add a [Tock port](https://github.com/apache/mynewt-nimble/tree/master/porting) to the Nimble code base. This is an example of the NuttX port: https://github.com/apache/mynewt-nimble/tree/master/porting/npl/nuttx/src

### Nimble hardware support

Nimble already supports two Tock boards the [Apollo3](https://github.com/apache/mynewt-nimble/tree/master/nimble/transport/apollo3) and the [nRF](https://github.com/apache/mynewt-nimble/tree/master/nimble/transport/nrf5340). Nimble directly access the hardware registers to send HCI commands to the radio.

This could be run directly in the kernel as is. Once Nimble is added as a capsule it will have hardware access. This can be used as a proof of concept to determine if the stack is working, without having to worry about hardware issues.

Then we would want to add a Tock HIL transport option to Nimble and use that instead. That way the hardware accesses can go from a capsule through a HIL to the chip radio.
